tests/system/src/test/java/test/robot/javafx/scene/RobotTest.java

```diff
@@ -0,0 +1,342 @@
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+package test.robot.javafx.scene;
+
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicReference;
+
+import javafx.application.Application;
+import javafx.application.Platform;
+import javafx.beans.InvalidationListener;
+import javafx.geometry.Point2D;
+import javafx.scene.Scene;
+import javafx.scene.control.Button;
+import javafx.scene.control.TextField;
+import javafx.scene.image.WritableImage;
+import javafx.scene.input.KeyCode;
+import javafx.scene.input.MouseButton;
+import javafx.scene.layout.Background;
+import javafx.scene.layout.BackgroundFill;
+import javafx.scene.layout.HBox;
+import javafx.scene.layout.Pane;
+import javafx.scene.layout.StackPane;
+import javafx.scene.paint.Color;
+import javafx.scene.robot.Robot;
+import javafx.stage.Screen;
+import javafx.stage.Stage;
+import javafx.stage.StageStyle;
+import javafx.stage.WindowEvent;
+
+import org.junit.AfterClass;
+import org.junit.BeforeClass;
+import org.junit.Test;
+import org.junit.Assert;
+import junit.framework.AssertionFailedError;
+
+import static javafx.scene.paint.Color.MAGENTA;
+import static org.junit.Assert.fail;
+
+/**
+ * Tests to verify that the native robot implementations all work correctly.
+ */
+public class RobotTest {
+    static CountDownLatch startupLatch = new CountDownLatch(1);
+    static volatile Stage stage;
+    static volatile Scene scene;
+    static Robot robot;
+    private static final double TOLERANCE = 0.07;
+    private static final int SIZE = 400;
+    private static final int WIDTH = 400;
+    private static final int HEIGHT = 300;
+
+    public static void main(String[] args) throws Exception {
+        RobotTest test = new RobotTest();
+        test.testKeyboard();
+        test.testMouseMove();
+        test.testMouseClick();
+        test.testMouseWheel();
+        test.testPixelCapture();
+        test.testScreenCapture();
+        exit();
+    }
+
+    @Test
+    public void testKeyboard() {
+        CountDownLatch onKeyTypedLatch = new CountDownLatch(1);
+        CountDownLatch setSceneLatch = new CountDownLatch(1);
+        TextField textField = new TextField();
+        InvalidationListener invalidationListener = observable -> setSceneLatch.countDown();
+        Platform.runLater(() -> {
```

Comment:
In system tests, Util.runAndWait is preferred to Platform.runLater since the former will propagate exceptions. It will also wait for the results, so that in some cases (but not all) you can eliminate the extra latch. In this case and in the case of waiting for specific events, there is still value in also waiting for those events to happen, so I don't recommend removing most of the latches.